Lib/asyncio/coroutines.py

```diff
@@ -189,15 +189,26 @@ def iscoroutine(obj):
 def _format_coroutine(coro):
     assert iscoroutine(coro)
 
-    if not hasattr(coro, 'cr_code') and not hasattr(coro, 'gi_code'):
-        # Most likely a built-in type or a Cython coroutine.
-
-        # Built-in types might not have __qualname__ or __name__.
-        coro_name = getattr(
-            coro, '__qualname__',
-            getattr(coro, '__name__', type(coro).__name__))
-        coro_name = f'{coro_name}()'
+    is_corowrapper = isinstance(coro, CoroWrapper)
+
+    def get_name(coro):
+        # Coroutines compiled with Cython sometimes don't have
+        # proper __qualname__ or __name__.  While that is a bug
```

Comment:
Do you have any insights into the conditions under which this can be observed? Is this a problem with older versions of Cython, or is there anything we can do to improve the situation?

Reply:
Not sure about `__qualname__` and `__name__`, I got reports from uvloop users about this a couple of months ago.  `__code__` set to `None` should be easily reproducible with latest Cython -- ideally, `__code__` should either be set to a code-like object or not set at all.